src/hotspot/cpu/aarch64/assembler_aarch64.hpp

```diff
@@ -2657,17 +2657,21 @@ void mvnw(Register Rd, Register Rm,
     f(sidx<<(int)T, 14, 11), f(1, 10), rf(Vn, 5), rf(Vd, 0);
   }
 
-#define INSN(NAME, op)                                                     \
-  void NAME(Register Rd, FloatRegister Vn, SIMD_RegVariant T, int idx) {   \
-    starti;                                                                \
-    f(0, 31), f(T==D ? 1:0, 30), f(0b001110000, 29, 21);                   \
-    f(((idx<<1)|1)<<(int)T, 20, 16), f(op, 15, 10);                        \
-    rf(Vn, 5), rf(Rd, 0);                                                  \
+  void umov(Register Rd, FloatRegister Vn, SIMD_RegVariant T, int idx) {
+    starti;
+    assert(T != Q, "invalid register variant");
+    f(0, 31), f(T == D ? 1 : 0, 30), f(0b001110000, 29, 21);
+    f(((idx << 1) | 1) << (int)T, 20, 16), f(0b001111, 15, 10);
+    rf(Vn, 5), rf(Rd, 0);
   }
 
-  INSN(umov, 0b001111);
-  INSN(smov, 0b001011);
-#undef INSN
+  void smov(Register Rd, FloatRegister Vn, SIMD_RegVariant T, int idx) {
+    starti;
+    assert(T < D, "invalid register variant");
+    f(0, 31), f(1, 30), f(0b001110000, 29, 21);
+    f(((idx << 1) | 1) << (int)T, 20, 16), f(0b001011, 15, 10);
+    rf(Vn, 5), rf(Rd, 0);
+  }
 
```

Comment:
I don't think the code duplication here is necessary. Please pass the validity test and the expression for Bit 30 into ```INSN```. It takes a few minutes studying the difference between those two in order to discover what the differences between them are, which could be saved if we made it explicit.

Reply:
Thanks for spotting that. Agree. Updated.